capa/features/extractors/ghidra/insn.py

```diff
@@ -0,0 +1,345 @@
+        # skip things like:
+        #   .text:0042250E retn 8
+        return
+
+    if capa.features.extractors.ghidra.helpers.is_sp_modified(insn):
+        # skip things like:
+        #   .text:00401145 add esp, 0Ch
+        return
+
+    for i in range(insn.getNumOperands()):
+        if insn.getOperandType(i) != OperandType.SCALAR:
+            # skip things like:
+            #   references, void types
+            continue
+
+        const = insn.getScalar(i).getValue()
+        addr = AbsoluteVirtualAddress(insn.getAddress().getOffset())
+
+        yield Number(const), addr
+        yield OperandNumber(i, const), addr
+
+        if insn.getMnemonicString().startswith("ADD") and 0 < const < MAX_STRUCTURE_SIZE:
+            # for pattern like:
+            #
+            #     add eax, 0x10
+            #
+            # assume 0x10 is also an offset (imagine eax is a pointer).
+            yield Offset(const), addr
+            yield OperandOffset(i, const), addr
+
+
+def extract_insn_offset_features(insn) -> Iterator[Tuple[Feature, Address]]:
+    """
+    parse instruction structure offset features
+
+    example:
+        .text:0040112F cmp [esi+4], ebx
+    """
+
+    # ignore any stack references
+    if not capa.features.extractors.ghidra.helpers.is_stack_referenced(insn):
+        # Ghidra stores operands in 2D arrays if they contain offsets
+        for i in range(insn.getNumOperands()):
+            if insn.getOperandType(i) == OperandType.DYNAMIC:  # e.g. [esi + 4]
+                # manual extraction, since the default api calls only work on the 1st dimension of the array
+                op_objs = insn.getOpObjects(i)
+                for j in range(len(op_objs)):
+                    if isinstance(op_objs[j], ghidra.program.model.scalar.Scalar):
+                        op_off = op_objs[j].getValue()
+                        yield Offset(op_off), AbsoluteVirtualAddress(insn.getAddress().getOffset())
+                        yield OperandOffset(i, op_off), AbsoluteVirtualAddress(insn.getAddress().getOffset())
+
+
+def extract_insn_bytes_features(insn) -> Iterator[Tuple[Feature, Address]]:
+    """
+    parse referenced byte sequences
+    example:
+        push    offset iid_004118d4_IShellLinkA ; riid
+    """
+
+    if insn.getMnemonicString().startswith("CALL"):
+        return
+
+    data_ref = OperandType.ADDRESS | OperandType.SCALAR  # DAT_* or s_*
+    ref = insn.getAddress()
+    for i in range(insn.getNumOperands()):
+        if insn.getOperandType(i) == data_ref:
+            ref = insn.getAddress(i)
+
+    if ref != insn.getAddress():
+        extracted_bytes = capa.features.extractors.ghidra.helpers.get_bytes(ref, MAX_BYTES_FEATURE_SIZE)
+        if extracted_bytes and not capa.features.extractors.helpers.all_zeros(extracted_bytes):
+            ghidra_dat = getDataAt(ref)  # type: ignore [name-defined] # noqa: F821
+            if ghidra_dat and not ghidra_dat.hasStringValue():
```

Comment:
Good call on the pre-check, thank you. I've gone ahead and moved it in the next commit. Also, the `getAddress(i)` line is what extracts a pointer if any. Otherwise, the function bails.

Reply:
@colton-gabertan that's true but we want to avoid pointers to external functions when extracting bytes e.g. run the following code snippet on `mimikatz`:
```python
for insn in currentProgram.getListing().getInstructions(True):
	for i in range(insn.getNumOperands()):
		if insn.getAddress(i):
			data = getDataAt(insn.getAddress(i))
			if data and data.isPointer():
				print(f"{insn.getAddress()} : {insn}")
```
You'll see many of these addresses are in fact pointers to external functions versus data we'd want to match using capa's `byte` feature.

Reply:
If you agree with my findings above then we should emit `ghidra_data.isPointer()`.